### PR TITLE
Add mpt lookups to state circuit

### PR DESCRIPTION
**Pull request by z2trillion**
Fixes https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/607

### CODE REVIEW

**Review by ed255 — COMMENTED** [review]
I've taken an initial look and everything looks correct.  I would like to review the spec part once it's ready and come back for a second look before approving.

On a second note, I'm a bit worried about debuggability in the lookups, since now we have the lookup expression being a single `Expression<F>`, and the lookup destination being a single column, where both side are RLC encoded.  So on failure, it's hard to figure out the original values of each column out of the RLC encoded value.  Does anyone share this worry?

On a high level I wonder if it would be possible to modify halo2 such that the lookups are kept as before (with all their original columns), and the RLC is performed by some halo2 functionality automatically (maybe only when working with the real prover).

File: zkevm-circuits/src/state_circuit/constraint_builder.rs [hunk 1]
```diff
@@ -129,6 +133,16 @@ impl<F: Field> ConstraintBuilder<F> {
                 q.initial_value.clone() - q.initial_value_prev(),
             );
         });
+
+        self.add_lookup(
+            "if state_root changes, mpt_update exists in mpt circuit",
+            (
+                q.lexicographic_ordering_selector.clone()
+                    * (q.state_root() - q.state_root_prev())
```
Comment (noel2004) [inline zkevm-circuits/src/state_circuit/constraint_builder.rs:141]: In my opinion, the lookup to mpt table should not be triggered by change of state_root, they should be triggered by some intrinsic rules (`Tag` is `AccountStorage` and it is the last row for a specified account / key, etc.). Following is the reasons:

1. the `state_root` col would be assigned arbitrarily in state circuit and there is no other way to provde them. For example we can assign identical value in `state_root` and all lookups are disabled.

2. Some entries are required in mpt table even it do not change state root. Consider a sole `SLOAD` to specified key or a `SELFBALANCE` op. We must provide it in mpt circuit but the state root after this op would never change.

Comment (z2trillion) [reply zkevm-circuits/src/state_circuit/constraint_builder.rs]: Thanks for catching this. I have changed the constraints to be what you suggested.

**Comment by z2trillion** [comment]
spec pr: https://github.com/privacy-scaling-explorations/zkevm-specs/pull/236

File: zkevm-circuits/src/state_circuit/constraint_builder.rs (hunk 1 above)
Comment (noel2004) [reply zkevm-circuits/src/state_circuit/constraint_builder.rs]: Great. I think now it is ok.

**Comment by ed255** [comment]
@z2trillion will this PR resolve https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/607 ?

**Comment by z2trillion** [comment]
yeah

**Pull request closed without merging**